### PR TITLE
[Admin] Promotions start and end date validation

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: features/promotion/managing_promotions/promotion_validation.feature [hunk 1]
```diff
@@ -27,10 +27,29 @@ Feature: Promotion validation
...
 
     @ui
+    Scenario: Adding a promotion with start date set up after ends date
+        Given I want to create a new promotion
+        When I specify its code as "FULL_METAL_PROMOTION"
+        And I name it "Full metal promotion"
+        And I specify that it should starts at "24.12.2017"
+        And I specify that it should ends at "12.12.2017"
+        And I try to add it
+        Then I should be notified that promotion cannot start after it ends
```
Comment (michalmarcinkowski) [inline features/promotion/managing_promotions/promotion_validation.feature]: IMO it is more natural to say that the promotion cannot end before it starts.

File: src/Sylius/Behat/Context/Transform/DateTimeContext.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+final class DateTimeContext implements Context
+{
+    /**
+     * @Transform :date
+     * @Transform :startsDate
+     * @Transform :endsDate
+     */
+    public function getPriceFromString($date)
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Transform/DateTimeContext.php]: `getPriceFromString` - not really 😆

File: src/Sylius/Bundle/PromotionBundle/Validator/PromotionDateRangeValidator.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+     * @param mixed $value
+     * @param Constraint $constraint
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        if (null === $value) {
+            return;
+        }
+
+        if (!$value instanceof PromotionInterface) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/PromotionBundle/Validator/PromotionDateRangeValidator.php]: `Assert::isInstanceOf`

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionsContext.php [hunk 4]
```diff
@@ -394,6 +394,57 @@ public function iShouldBeNotifiedOfFailure()
     }
 
     /**
+     * @When I specify that it should starts at :date
+     */
+    public function iSpecifyThatItShouldStartAt(\DateTime $date)
+    {
+        $currentPage = $this->currentPageResolver->getCurrentPageWithForm($this->createPage, $this->updatePage);
+
+        $currentPage->setStarsAt($date);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionsContext.php]: $currentPage->setStar**t**sAt($date);

File: src/Sylius/Bundle/PromotionBundle/Validator/PromotionDateRangeValidator.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    public function validate($value, Constraint $constraint)
+    {
+        if (null === $value) {
+            return;
+        }
+
+        Assert::isInstanceOf(
+            $value,
+            PromotionInterface::class,
+            sprintf('Expecting %s class. %s given', PromotionInterface::class, get_class($value))
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/PromotionBundle/Validator/PromotionDateRangeValidator.php]: The default message is `'Expected an instance of %2$s. Got: %s'`, so this custom one is not needed.

File: features/promotion/managing_promotions/promotion_validation.feature [hunk 6]
```diff
@@ -27,10 +27,29 @@ Feature: Promotion validation
...
         When I remove its name
         And I try to save my changes
         Then I should be notified that name is required
         And this promotion should still be named "Christmas sale"
+
+    @ui
+    Scenario: Trying to add start later then ends date for existing promotion
+        Given there is a promotion "Christmas sale"
+        And I want to modify this promotion
+        And I specify that it should starts at "24.12.2017"
```
Comment (pjedrzejewski) [inline features/promotion/managing_promotions/promotion_validation.feature]: start~~s~~ here and everywhere else.

**Comment by michalmarcinkowski** [comment]
Thanks Łukasz! 👍

**Pull request merged**